### PR TITLE
[TS] Add TypeScript unit testing support

### PR DESCRIPTION
**Pull request by alexmirrington**
#### What does this PR do?

This is a proof-of-concept PR related to #5111 that adds **strict** TypeScript unit testing support, without forcing the rest of the `grommet` code base to use strict type checks. This means that we can gradually convert existing unit tests from `*-test.js` to `*-test.tsx` files, and fix any issues that TypeScript raises in the process. This should allow us to find and fix existing TypeScript issues quicker, as well as decrease opportunities for TS-related bugs in the future.

#### Where should the reviewer start?

The most important change is in `package.json`, where I specify we want to use the new `ts-jest` dependency for all `tsx?` tests, but use existing pipelines (`babel-jest`) for all `.jsx?` tests.

Secondly, there have been some changes to the `tsconfig.json` file, to support the creation of a more strict `tsconfig.tests.json` file that only applies to files in `**/__tests__` directories. The removed fields are to allow the `strict: true` field in the `tsconfig.tests.json` file, and are all consistent with `tsconfig` [defaults](https://www.typescriptlang.org/tsconfig), meaning type strictness has not been affected for the main `grommet` codebase, only tests. The reviewer should double check that all removed fields are consistent with `tsconfig` [defaults](https://www.typescriptlang.org/tsconfig).

#### What testing has been done on this PR?

All existing unit tests still pass despite the newly added TypeScript support, both in my local `pre-commit` hook and on CI. See manual testing below for more details.

#### How should this be manually tested?

The reviewer should test that converting a `.js` test file for a few components to `.tsx` should cause tests to fail, since a bunch of new strict checks have been put in place. Failures may range from implicit `any` types through to grommet bugs caused my missing/incorrect prop types. In my testing, I used the `Accordion` and `DataTable` components, however I'd suggest the reviewer tries a few other components too. You can also change some correct prop interfaces to check that errors are correctly identified both in your code editor and when running `yarn test`. In some cases such as the `Anchor` component, there are not type errors at all, and conversion to typescript is as simple as renaming the test file from `.js` to `.tsx` and updating the test spec.

#### Any background context you want to provide?

All context is described here: #5111
This PR is an updated version of #5127, as we have had a few major upgrades to our dependencies since then.

#### Do the grommet docs need to be updated? Should this PR be mentioned in the release notes?

No, it is internal to Grommet.

#### Is this change backwards compatible or is it a breaking change?

Backwards compatible

### CODE REVIEW

File: tsconfig.json [hunk 1]
```diff
@@ -6,23 +6,20 @@
     "moduleResolution": "node",
     "outDir": "./dist/",
     "sourceMap": true,
-    "noImplicitAny": false,
     "module": "esnext",
     "target": "es6",
     "jsx": "react",
-    "strictPropertyInitialization": false,
     "declaration": true,
-    "strict": false,
```
Comment (jcfilben) [inline tsconfig.json:15]: I'm curious why we are removing these settings

File: tsconfig.tests.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "extends": "./tsconfig.json",
+  "compilerOptions": {
+    "strict": true,
+    "noUnusedLocals": true,
+    "noUnusedParameters": true,
+    "noImplicitReturns": true,
+    "noFallthroughCasesInSwitch": true
```
Comment (jcfilben) [inline tsconfig.tests.json:8]: Why these specific options, are they just rules you think should be enforced or is it a recommended pattern?

**Review by jcfilben — COMMENTED** [review]
Nice job! I converted some of the test files to .tsx files and looks like the js tests and tsx tests are running as expected. I just had a question about the tsconfig files

File: tsconfig.json (hunk 1 above)
Comment (alexmirrington) [reply tsconfig.json:15]: There are two reasons:
- These settings are the defaults for `.tsconfig` files, so removing them has not impact on the strictness of the Typescript checks used for the main part of the repo
- When extending `.tsconfig.yaml` in `.tsconfig.tests.yaml`, it's important to note that only ["files, include and exclude from the inheriting config file overwrite those from the base config file"](https://www.typescriptlang.org/tsconfig#extends). What this means for us is that including `strict`, `noUnusedLocals`, `noUnusedParameters` etc. in the base `tsconfig` stops us from overriding these values in the extended `tsconfig`, contrary to the intuition we usually have from OOP.

File: tsconfig.tests.json (hunk 2 above)
Comment (alexmirrington) [reply tsconfig.tests.json:8]: We could definitely add more to be more strict, but these were some that I have found to be useful in some of my own TypeScript projects. If I recall correctly they were adapted from a default strict tsconfig. Here's what they do:

```
    /* Strict Type-Checking Options */
    "strict": true /* Enable all strict type-checking options: noImplicitAny, strictNullChecks, strictFunctionTypes, strictPropertyInitialization, noImplicitThis, alwaysStrict */,
    "noUnusedLocals": true /* Report errors on unused locals. */,
    "noUnusedParameters": true /* Report errors on unused parameters. Requires TypeScript version 2.0 or later.*/,
    "noImplicitReturns": true /* Report error when not all code paths in function return a value. */,
    "noFallthroughCasesInSwitch": true /* Report errors for fallthrough cases in switch statement. */,
```

File: tsconfig.json (hunk 1 above)
Comment (jcfilben) [reply tsconfig.json:15]: Thank you for the detailed explanation!

File: tsconfig.tests.json (hunk 2 above)
Comment (jcfilben) [reply tsconfig.tests.json:8]: Great, thanks for providing detail for each rule

**Review by jcfilben — APPROVED** [review]
Looks good to me

**Comment by alexmirrington** [comment]
FYI: The force push was to resolve `yarn.lock` locally and rebase onto the current master.

**Review by ericsoderberghp — APPROVED** [review]
I'm good with this as a building block to start with. I like how the Anchor test needs no changes. It will be interesting to see how large the changes are elsewhere as we go.

**Comment by ericsoderberghp** [comment]
BTW, thanks for an excellent Description for this pull request.

**Comment by alexmirrington** [comment]
@ericsoderberghp You're welcome! I'll create a separate issue to start tracking test migration progress. I'll save most of the discussion for that ticket, however the way I would recommend we start is to try converting tests to `.tsx` at first, and if there are a small amount of TS errors they can be fixed in the same PR as the test migration, otherwise separate issues can be created prior to migrating the test files to avoid massive PRs.

**Pull request merged**